File: pythonX/medium/m018_clone_graph.py
"""
M018: Clone Graph (Deep Copy)

Summary:
Create a deep copy of an undirected graph.

Requirements:
Given a reference to a node, return deep copy of graph.

Approach:
Use DFS with dict to track cloned nodes.

Data Structures:
- Node class with val and neighbors list
- dict[Node, Node]: original -> clone mapping

Big O Analysis:
Time Complexity: O(V + E)
Space Complexity: O(V)
"""

from typing import Optional


class Node:
    def __init__(self, val: int = 0, neighbors: Optional[list['Node']] = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


def clone_graph(node: Optional[Node]) -> Optional[Node]:
    """
    TODO: Clone graph
    1. If node is None, return None
    2. Create dict for tracking clones
    3. Call helper DFS function
    """
    pass


def clone(node: Node, visited: dict[Node, Node]) -> Node:
    """
    TODO: Recursive clone helper
    1. If node in visited, return cloned node
    2. Create new node with same value
    3. Add to visited dict
    4. For each neighbor:
       - Recursively clone neighbor
       - Add to new node's neighbors
    5. Return new node
    """
    pass


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Four-node graph")
    # Graph: 1--2
    #        |  |
    #        4--3
    node1 = Node(1)
    node2 = Node(2)
    node3 = Node(3)
    node4 = Node(4)
    node1.neighbors = [node2, node4]
    node2.neighbors = [node1, node3]
    node3.neighbors = [node2, node4]
    node4.neighbors = [node1, node3]
    
    cloned1 = clone_graph(node1)
    print(f"Cloned graph node value: {cloned1.val if cloned1 else 'null'}")
    print(f"Is different object: {cloned1 is not node1}")
    print(f"Has same neighbors count: {len(cloned1.neighbors) == 2 if cloned1 else False}")
    
    print("\nTest 2: Single node")
    single = Node(1)
    cloned_single = clone_graph(single)
    print(f"Cloned single node: {cloned_single.val if cloned_single else 'null'}")
    
    print("\nTest 3: Null graph")
    cloned_null = clone_graph(None)
    print(f"Cloned null: {'null' if cloned_null is None else 'not null'}")

File: src/com/example/javas/medium/M018_CloneGraph.java
/*
MEDIUM PROBLEM (Simplified for 25 min):
Clone Graph (Deep Copy of Graph)

REQUIREMENTS:
- Given a reference to a node in a connected undirected graph
- Return a deep copy (clone) of the graph
- Each node contains a value and a list of neighbors

APPROACH:
- Use DFS with HashMap to track cloned nodes
- HashMap maps original node -> cloned node
- Recursively clone all neighbors

DATA STRUCTURES:
- HashMap<Node, Node>: original -> clone mapping
- Node class with val and neighbors list

TIME COMPLEXITY: O(V + E) where V = nodes, E = edges
SPACE COMPLEXITY: O(V)
*/

package com.example.javas.medium;

import java.util.*;

public class M018_CloneGraph {
    
    // Node definition
    static class Node {
        public int val;
        public List<Node> neighbors;
        
        public Node(int val) {
            this.val = val;
            this.neighbors = new ArrayList<>();
        }
    }
    
    public static Node cloneGraph(Node node) {
        // TODO: Implement graph cloning
        // 1. If node is null, return null
        // 2. Create HashMap for tracking cloned nodes
        // 3. Call helper DFS function
        
        return null;
    }
    
    private static Node clone(Node node, HashMap<Node, Node> visited) {
        // TODO: Implement recursive clone helper
        // 1. If node already in visited map, return cloned node
        // 2. Create new node with same value
        // 3. Add to visited map
        // 4. For each neighbor:
        //    - Recursively clone neighbor
        //    - Add cloned neighbor to new node's neighbors list
        // 5. Return new node
        
        return null;
    }
    
    // TEST CASES
    public static void main(String[] args) {
        // Graph 1: 1--2
        //          |  |
        //          4--3
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        node1.neighbors.add(node2);
        node1.neighbors.add(node4);
        node2.neighbors.add(node1);
        node2.neighbors.add(node3);
        node3.neighbors.add(node2);
        node3.neighbors.add(node4);
        node4.neighbors.add(node1);
        node4.neighbors.add(node3);
        
        Node cloned1 = cloneGraph(node1);
        System.out.println("Cloned graph 1, node value: " + (cloned1 != null ? cloned1.val : "null"));
        System.out.println("Is different object: " + (cloned1 != node1));
        System.out.println("Has same neighbors count: " + (cloned1 != null && cloned1.neighbors.size() == 2));
        
        // Graph 2: single node
        Node single = new Node(1);
        Node clonedSingle = cloneGraph(single);
        System.out.println("\nCloned single node: " + (clonedSingle != null ? clonedSingle.val : "null"));
        
        // Graph 3: null
        Node clonedNull = cloneGraph(null);
        System.out.println("Cloned null: " + (clonedNull == null ? "null" : "not null"));
    }
}
